pkg/persistence/object_store.go

```diff
@@ -215,40 +218,28 @@ func (s *objectBackupStore) PutBackup(info BackupInfo) error {
 		return kerrors.NewAggregate([]error{err, deleteErr})
 	}
 
-	if err := seekAndPutObject(s.objectStore, s.bucket, s.layout.getPodVolumeBackupsKey(info.Name), info.PodVolumeBackups); err != nil {
-		errs := []error{err}
-
-		deleteErr := s.objectStore.DeleteObject(s.bucket, s.layout.getBackupContentsKey(info.Name))
-		errs = append(errs, deleteErr)
-
-		deleteErr = s.objectStore.DeleteObject(s.bucket, s.layout.getBackupMetadataKey(info.Name))
-		errs = append(errs, deleteErr)
-
-		return kerrors.NewAggregate(errs)
-	}
-
-	if err := seekAndPutObject(s.objectStore, s.bucket, s.layout.getBackupVolumeSnapshotsKey(info.Name), info.VolumeSnapshots); err != nil {
-		errs := []error{err}
-
-		deleteErr := s.objectStore.DeleteObject(s.bucket, s.layout.getBackupContentsKey(info.Name))
-		errs = append(errs, deleteErr)
-
-		deleteErr = s.objectStore.DeleteObject(s.bucket, s.layout.getBackupMetadataKey(info.Name))
-		errs = append(errs, deleteErr)
-
-		return kerrors.NewAggregate(errs)
+	// Since the logic for all of these files is the exact same except for the name and the contents,
+	// use a map literal to iterate through them and write them to the bucket.
+	var backupObjs = map[string]io.Reader{
+		s.layout.getPodVolumeBackupsKey(info.Name):       info.PodVolumeBackups,
+		s.layout.getBackupVolumeSnapshotsKey(info.Name):  info.VolumeSnapshots,
+		s.layout.getBackupResourceListKey(info.Name):     info.BackupResourceList,
+		s.layout.getCSIVolumeSnapshotKey(info.Name):      info.CSIVolumeSnapshots,
```

Comment:
I am thinking if we should use this to get the CSI snapshots to delete instead of what I am doing here https://github.com/vmware-tanzu/velero/pull/2411/files#diff-afe709fa1785e08707ca0825e3f60e68R474
This would be consistent w/ how we discover and delete non-csi volumesnapshots

Reply:
That would unify the approaches for sure. I think this goes to @skriss's questions about synchronization - which is the source of truth on the CSI snapshots, the object storage, or the kube API server?

Historically, the object storage has been Velero's source of truth. It may make sense to keep going with that approach for these objects.

Before we can rely on it completely, we need to ensure that Volume Snapshot Contents objects are labeled with the backup's name at backup time, as they currently are not, so the file's always empty now.

Reply:
Also, for what it's worth, when deleting a Velero backup, I think we'll need to delete objects in the cluster and from object storage, regardless of which is the source of truth. But deleting them from object storage "should" be as simple as removing the backup's root key, which would then take the rest of the contents.

Reply:
Yeah, deleting the `.json.gz` files from object storage should already be handled by this struct's `DeleteBackup` func which deletes everything under the backup prefix.

@ashish-amarnath the issue is that if the VolumeSnapshot/VolumeSnapshotContents don't exist in the cluster, then there's nothing *to* delete -- with native Velero snapshots, we delete the storage snapshot directly via the cloud provider API, but with CSI, all we can delete is the in-cluster custom resources.

Reply:
I think we've gotten what we want from this conversation! marking this as resolved.